src/main/java/seedu/duke/TaskList.java

```diff
@@ -0,0 +1,235 @@
+     * @param storage The hard disk for storage of data.
+     */
+    public TaskList(List<Task> tasks, Storage storage) {
+        this.tasks = tasks;
+        this.storage = storage;
+        ui = new Ui();
+    }
+
+    protected List<Task> getTasks() {
+        return tasks;
+    }
+    /**
+     * Adds the user input of todo task into the task list.
+     *
+     * @param desc The details of the todo task.
+     * @param doneStatus An indicator which shows whether a todo task has been completed or not.
+     * @throws IOException If an input or output exception occurred.
+     */
+    protected void addTodo(String desc, String doneStatus) throws IOException {
+        Task todo = new Todo(desc);
+        if (doneStatus.equalsIgnoreCase("Y")) {
+            todo.markAsDone();
+        }
+        tasks.add(todo);
+        storage.addToStorage(todo);
+        printAddToList();
+        ui.print(todo.toString());
+        printNumTask();
+    }
+
+    /**
+     * Adds the user input of deadline task into the task list.
+     *
+     * @param desc The details of the deadline task.
+     * @param doneStatus An indicator which shows whether a deadline task has been completed or not.
+     * @throws InvalidTaskInputException If an invalid task command is input.
+     * @throws IOException If an input or output exception occurred.
+     * @throws InvalidDateException If a date is input in a wrong format.
+     */
+    protected void addDeadline(String desc, String doneStatus)
+            throws InvalidTaskInputException, IOException, InvalidDateException {
+        String[] descs = desc.split(" /by |\\|");
+        if (descs.length == 1) { // invalid Deadline input format
+            throw new InvalidTaskInputException();
+        }
+
+        String deadlineDesc = descs[0].trim();
+        String deadlineTime = descs[1].trim();
+        LocalDate formattedDeadlineTime = null;
+        if (isValidDate(deadlineTime)) {
+            formattedDeadlineTime = LocalDate.parse(deadlineTime);
+        } else {
+            throw new InvalidDateException();
+        }
+        Task deadline = new Deadline(deadlineDesc, formattedDeadlineTime);
+
+        if (doneStatus.equalsIgnoreCase("Y")) {
+            deadline.markAsDone();
+        }
+        tasks.add(deadline);
+        storage.addToStorage(deadline);
+        printAddToList();
+        ui.print(deadline.toString());
+        printNumTask();
+    }
+
+    /**
+     * Adds the user input of event task into the task list.
+     *
+     * @param desc The details of the event task.
+     * @param doneStatus An indicator which shows whether an event task has been completed or not.
+     * @throws InvalidTaskInputException If an invalid task command is input.
+     * @throws IOException If an input or output exception occurred.
+     * @throws InvalidDateException If a date is input in a wrong format.
+     */
+    protected void addEvent(String desc, String doneStatus)
```

Comment:
consider abstracting into a Task class so you can rely on Polymorphism to add the Tasks!

Reply:
you're correct, thank you for the suggestion!